fix: restore AIO on systems with larger min size

mnitchev opened the pull request:
The AIO size has a minimum based on amount of CPUs available. If the
size that we dumped is smaller than the minimum on the destination we
need to create a ring in a new location and fix the references to it in
the VMAs.
This patch of the references isn't ideal, but as this causes the
restoration to fail, it is best effort.

apasyniuk commented on criu/pie/restorer.c:
```diff
@@ -921,12 +921,76 @@ static unsigned long restore_mapping(VmaEntry *vma_entry)
+		 * Scan this VMA for the old address value.
+		 * Only check aligned positions (aio_context_t is unsigned long).
+		 */
+		for (addr = vma->start; addr + sizeof(unsigned long) <= vma->end;
```
Could you check if this will be vectorised? It's a simple compare-and-swap operation, but it consumes a lot of memory if the process has a lot of it. So it would be nice if it is optimized by compilator.

kr3v replied:
nit: https://godbolt.org/z/Yr976b1Wj

but, per claude:
> ### Practical Reality
> Given this appears to be kernel code scanning VMAs, the bottleneck is likely memory access latency, not compute. Vectorization won't help much if you're memory-bound with unpredictable page faults. The compiler might not auto-vectorize due to the pointer aliasing and side effects.
Verdict: Probably not worth the complexity unless profiling shows this specific loop is a major bottleneck.

mnitchev replied:
So can I resolve this then? 😄

apasyniuk replied:
Plus, for profiling or just checking how long it takes.
However, TBH the rest of the claude analysis is kind of gibberish :D

kr3v replied:
> Plus, for profiling or just checking how long it takes. However, TBH the rest of the claude analysis is kind of gibberish :D

my main point was - "the bottleneck is likely memory access latency, not compute"
that we'd basically be blocked on waiting memory pages to be fetched to CPU cache

apasyniuk commented on criu/pie/restorer.c:
```diff
@@ -921,12 +921,76 @@ static unsigned long restore_mapping(VmaEntry *vma_entry)
+		 * Only check aligned positions (aio_context_t is unsigned long).
+		 */
+		for (addr = vma->start; addr + sizeof(unsigned long) <= vma->end;
+		     addr += sizeof(unsigned long)) {
```
Something is of with the indentatioon.

mnitchev replied:
it just decides that the line is too long, but should be otherwise okay.

apasyniuk commented on criu/pie/restorer.c:
```diff
@@ -946,14 +1012,41 @@ static int restore_aio_ring(struct rst_aio_ring *raio)
+	needs_relocation = (new_ring_len > raio->len);
+	raio->new_addr = 0;
+
+	if (new->nr > ring->nr) {
```
Coudn't we check if this will actaully fit into the place that is is remaped into, instead always assuming that relocation is needed?

mnitchev replied:
I have added this, but the our golang app always puts a VMA adjacent to the AIO, so I can't really test this.

mnitchev committed:
fix: restore AIO on systems with larger min size

The AIO size has a minimum based on amount of CPUs available. If the
size that we dumped is smaller than the minimum on the destination we
need to create a ring in a new location and fix the references to it in
the VMAs. This patch of the references isn't ideal, but as this causes the
restoration to fail, it is best effort.
